test/unit/datetime.c

```diff
@@ -0,0 +1,261 @@
+	dt_to_struct_tm(local_dt(dt), &tm);
+
+	int seconds_of_day = (int64_t)dt->epoch % 86400;
+	tm.tm_hour = (seconds_of_day / 3600) % 24;
+	tm.tm_min = (seconds_of_day / 60) % 60;
+	tm.tm_sec = seconds_of_day % 60;
+
+	return &tm;
+}
+
+static void datetime_test(void)
+{
+	size_t index;
+	int64_t secs_expected;
+	int32_t nanosecs;
+	int32_t offset;
+
+	plan(355);
+	parse_datetime(sample, sizeof(sample) - 1,
+		       &secs_expected, &nanosecs, &offset);
+
+	for (index = 0; index < lengthof(tests); index++) {
+		int64_t secs;
+		int rc = parse_datetime(tests[index].str, tests[index].len,
+					&secs, &nanosecs, &offset);
+		is(rc, 0, "correct parse_datetime return value for '%s'",
+		   tests[index].str);
+		is(secs, secs_expected, "correct parse_datetime output "
+					"seconds for '%s",
+		   tests[index].str);
+
+		/*
+		 * check that stringized literal produces the same date
+		 * time fields
+		 */
+		static char buff[40];
+		struct datetime dt = {secs, nanosecs, offset, 0};
+		/* datetime_to_tm returns time in GMT zone */
+		struct tm *p_tm = datetime_to_tm(&dt);
+		size_t len = strftime(buff, sizeof(buff), "%F %T", p_tm);
+		ok(len > 0, "strftime");
+		int64_t parsed_secs;
+		int32_t parsed_nsecs, parsed_ofs;
+		rc = parse_datetime(buff, len, &parsed_secs, &parsed_nsecs, &parsed_ofs);
+		is(rc, 0, "correct parse_datetime return value for '%s'", buff);
+		is(secs, parsed_secs,
+		   "reversible seconds via strftime for '%s", buff);
+	}
+	check_plan();
+}
+
+
+static void
+tostring_datetime_test(void)
+{
+	static struct {
+		const char *string;
+		int64_t     secs;
+		uint32_t    nsec;
+		uint32_t    offset;
+	} tests[] = {
+		{"1970-01-01T02:00+02:00",          0,         0,  120},
+		{"1970-01-01T01:30+01:30",          0,         0,   90},
+		{"1970-01-01T01:00+01:00",          0,         0,   60},
+		{"1970-01-01T00:01+00:01",          0,         0,    1},
+		{"1970-01-01T00:00Z",               0,         0,    0},
+		{"1969-12-31T23:59-00:01",          0,         0,   -1},
+		{"1969-12-31T23:00-01:00",          0,         0,  -60},
+		{"1969-12-31T22:30-01:30",          0,         0,  -90},
+		{"1969-12-31T22:00-02:00",          0,         0, -120},
+		{"1970-01-01T00:00:00.123456789Z",  0, 123456789,    0},
+		{"1970-01-01T00:00:00.123456Z",     0, 123456000,    0},
+		{"1970-01-01T00:00:00.123Z",        0, 123000000,    0},
+		{"1973-11-29T21:33:09Z",    123456789,         0,    0},
+		{"2013-10-28T17:51:56Z",   1382982716,         0,    0},
+		{"9999-12-31T23:59:59Z", 253402300799,         0,    0},
+	};
+	size_t index;
+
+	plan(15);
+	for (index = 0; index < lengthof(tests); index++) {
+		struct datetime date = {
+			tests[index].secs,
+			tests[index].nsec,
+			tests[index].offset,
+			0
+		};
+		char buf[48];
+		datetime_to_string(buf, sizeof(buf), &date);
+		is(strcmp(buf, tests[index].string), 0,
+		   "string '%s' expected, received '%s'",
+		   tests[index].string, buf);
+	}
+	check_plan();
+}
+
+int
+main(void)
+{
+	plan(2);
```

Comment:
Please, use header() and footer() functions. Globally for the test file, and for each of the sub-suites with their individual plans.

Reply:
I'm curious - what's the big purpose of `header()`/`footer()` functions? I looked around, and, apparently, we do not use them consistently: there are plenty of tests which do not call header/footer, some call it in every function, some call only once in the main?

Result of their usage looks ugly, not very TAP way, and usefulness of them are usually close to zero (because subtest will be separately titled in any case).

Reply:
> what's the big purpose of header()/footer() functions?

Make output more readable.

> we do not use them consistently

It does not mean that more inconsistency must be added.

> there are plenty of tests which do not call header/footer, some call it in every function, some call only once in the main?

These are bad. Some are very old done before anybody cared about the code being clean, a few new have been merged without a proper review.

> Result of their usage looks ugly,

These header/footer prints help to separate test cases from each other. IMO the monolithic walls of indistinguishable `ok`s are ugly.

> not very TAP way

They are printed according to TAP, as comments.

> usefulness of them are usually close to zero 

This is not true really. As I said, they help to separate test case blocks from each other. Also they help to reduce diff significantly when some test case blocks get new cases. Then it won't affect numerations and plans of other blocks.